File: frontend/src/lib/signing.ts
function canonicalJson(obj: unknown, sortedKeys?: string[]): string {
  if (obj === null || obj === undefined) return 'null';
  if (typeof obj === 'string') return JSON.stringify(obj);
  if (typeof obj === 'number' || typeof obj === 'boolean') return String(obj);
  if (Array.isArray(obj)) {
    const items = obj.map((v) => canonicalJson(v));
    return '[' + items.join(',') + ']';
  }
  if (typeof obj === 'object') {
    const keys = sortedKeys ?? Object.keys(obj as Record<string, unknown>).sort();
    const pairs = keys
      .filter((k) => (obj as Record<string, unknown>)[k] !== undefined)
      .map((k) => JSON.stringify(k) + ':' + canonicalJson((obj as Record<string, unknown>)[k]));
    return '{' + pairs.join(',') + '}';
  }
  return String(obj);
}

async function computeSHA256(data: string): Promise<string> {
  const encoder = new TextEncoder();
  const bytes = encoder.encode(data);
  const hashBuffer = await crypto.subtle.digest('SHA-256', bytes);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  return hashArray.map((b) => b.toString(16).padStart(2, '0')).join('');
}

function toISO(t: number | null): string | null {
  return t ? new Date(t).toISOString() : null;
}

export async function computeSessionHash(session: {
  id: string;
  start: number | null;
  end: number | null;
  mode: string;
  finalScore: number | null;
  pctFocused: number | null;
  verdict: string | null;
  events: Array<{
    eventType: string;
    timestampS: number;
    confidence: number | null;
  }>;
}): Promise<string> {
  const eventsSorted = (session.events ?? [])
    .map((e) => ({
      event_type: e.eventType,
      timestamp_s: e.timestampS,
      confidence: e.confidence,
    }))
    .sort((a, b) => {
      const tDiff = a.timestamp_s - b.timestamp_s;
      return tDiff !== 0 ? tDiff : a.event_type.localeCompare(b.event_type);
    });

  const payload: Record<string, unknown> = {
    session_id: session.id,
    start: toISO(session.start),
    end: toISO(session.end),
    mode: session.mode,
    final_score: session.finalScore,
    pct_focused: session.pctFocused,
    verdict: session.verdict,
    events: eventsSorted,
  };

  const sortedKeys = Object.keys(payload).sort();
  const canonical = canonicalJson(payload, sortedKeys);
  return computeSHA256(canonical);
}

interface VerifyResponse {
  valid: boolean;
}

export async function verifySessionHash(
  sessionId: string,
  signature: string,
): Promise<boolean> {
  const baseUrl = import.meta.env.VITE_API_URL ?? '';
  try {
    const res = await fetch(`${baseUrl}/api/verify`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ session_id: sessionId, signature }),
    });
    if (!res.ok) return false;
    const data: VerifyResponse = await res.json();
    return data.valid;
  } catch {
    return false;
  }
}

export async function fetchSessionHash(
  sessionId: string,
): Promise<string | null> {
  const baseUrl = import.meta.env.VITE_API_URL ?? '';
  try {
    const res = await fetch(`${baseUrl}/api/verify/${sessionId}`);
    if (!res.ok) return null;
    const data = await res.json();
    return data.hash as string;
  } catch {
    return null;
  }
}

File: backend/report/signing.py
from __future__ import annotations

import hashlib
import json
from typing import Any

from backend.models.session import Session


def _canonical_json(obj: Any, *, sort_keys: bool = True) -> str:
    return json.dumps(obj, sort_keys=sort_keys, separators=(",", ":"), ensure_ascii=False)


def sign_session(session: Session) -> str:
    events_sorted = sorted(
        [
            {
                "event_type": e.event_type,
                "timestamp_s": e.timestamp_s,
                "confidence": e.confidence,
            }
            for e in session.events
        ],
        key=_sort_key_for_event,
    )

    payload = {
        "session_id": session.id,
        "start": session.start.isoformat() if session.start else None,
        "end": session.end.isoformat() if session.end else None,
        "mode": session.mode,
        "final_score": session.final_score,
        "pct_focused": session.pct_focused,
        "verdict": session.verdict.value if session.verdict else None,
        "events": events_sorted,
    }

    canonical = _canonical_json(payload)
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def _sort_key_for_event(e: dict[str, Any]) -> tuple[float, str]:
    return (float(e["timestamp_s"]), str(e["event_type"]))


def verify_signature(session: Session, signature: str) -> bool:
    expected = sign_session(session)
    return _constant_time_compare(expected, signature)


def _constant_time_compare(a: str, b: str) -> bool:
    if len(a) != len(b):
        return False
    result = 0
    for ca, cb in zip(a.encode("utf-8"), b.encode("utf-8"), strict=True):
        result |= ca ^ cb
    return result == 0
